<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Nick Strayer">
<meta name="dcterms.date" content="2019-10-16">
<meta name="description" content="This post is a longer-form and less-formal accompaniment to the manuscript “PheWAS-ME: A web-app for interactive exploration of multimorbidity patterns in PheWAS” and accompanying application. As the first of three papers that make up my PhD dissertation, the project represents a significant collaborative effort bringing together Electronic Health Records (EHR) and Biobank data using R and Shiny.">

<title>- PheWAS-ME, an app for exploration of multimorbidity patterns in PheWAS</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

<script type="text/javascript">

(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
ga('create', 'UA-92165973-1', 'auto');

ga('send', {
  hitType: 'pageview',
  'anonymizeIp': true,
});
</script>
<script src="https://code.jquery.com/jquery-3.2.1.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/d3/4.9.1/d3.min.js"></script>

<style>
  #custom-header {
    height: 30vh;
    width: 100%;
    position:relative;
    z-index:1;
    margin:0 auto;
  }
</style>


<meta property="og:title" content="- PheWAS-ME, an app for exploration of multimorbidity patterns in PheWAS">
<meta property="og:description" content="This post is a longer-form and less-formal accompaniment to the manuscript “PheWAS-ME: A web-app for interactive exploration of multimorbidity patterns in PheWAS” and accompanying application. As the first of three papers that make up my PhD dissertation, the project represents a significant collaborative effort bringing together Electronic Health Records (EHR) and Biobank data using R and Shiny.">
<meta property="og:image" content="https://www.livefreeordichotomize.com/media/phewas_me/whole_app.png">
<meta property="og:site-name" content=" ">
<meta property="og:image:height" content="2470">
<meta property="og:image:width" content="4290">
<meta name="twitter:title" content="Live Free or Dichotomize">
<meta name="twitter:description" content="This post is a longer-form and less-formal accompaniment to the manuscript “PheWAS-ME: A web-app for interactive exploration of multimorbidity patterns in PheWAS” and accompanying application. As the first of three papers that make up my PhD dissertation, the project represents a significant collaborative effort bringing together Electronic Health Records (EHR) and Biobank data using R and Shiny.">
<meta name="twitter:image" content="https://www.livefreeordichotomize.com/posts/2019-10-16-phewas_me/media/lfod-card.png">
<meta name="twitter:creator" content="@LucyStats">
<meta name="twitter:card" content="summary_large_image">
</head><body class="nav-fixed fullcontent"><header id="custom-header">
  <div id="header_viz"></div>
</header>

<script>
  $( document ).ready(() => {

    //javascript for intro loader

    //Function to immitate R's seq
    function seq(start, end, increment) {
      var length = Math.round((end - start)/increment);
      return Array(length).fill().map((_, idx) => start + idx*increment)
    }


    //define the pdf of the distribution.
    var logistic = function(x, theta, i) {
        var mu = 0.1;
        sign = 1
        var y = sign * (1 / (Math.sqrt(2 * Math.PI) * theta)) * (1 / x) *
            Math.exp(-Math.pow((Math.log(x) - mu), 2) / (2 * Math.pow(theta, 2)))
        return y;
    }

    var width   = $("#custom-header").width() ,
        height  = $("#custom-header").outerHeight(),
        padding = 0,
        numOfLines = 20,
        xs = seq(0.01, 5, .01),
        colors = ['rgb(165,0,38)', 'rgb(215,48,39)', 'rgb(244,109,67)', 'rgb(253,174,97)', 'rgb(254,224,144)',
            'rgb(224,243,248)', 'rgb(171,217,233)', 'rgb(116,173,209)', 'rgb(69,117,180)', 'rgb(49,54,149)'
        ];
  
    console.log("trying to run this shit.")
    //define the svg.
    var svg = d3.select("#header_viz").append("svg")
        .attr("width", width)
        .attr("height", height + 2 * padding)
        .append("g")


    var animatelines = function(whichline) {
        d3.selectAll(".line").style("opacity","0.5");

        //Select All of the lines and process them one by one
        d3.selectAll(".line").each(function(d,i){
            // Get the length of each line in turn
            var totalLength = d3.select("#line" + i).node().getTotalLength();

            d3.selectAll("#line" + i).attr("stroke-dasharray", totalLength + " " + totalLength)
              .attr("stroke-dashoffset", totalLength)
              .transition()
              .duration(5000)
              .delay(100*i)
              .attr("stroke-dashoffset", 0)
              .style("stroke-width",2)
        })

        writeGreeting()
    }



    // The Scales:
    var thetaMap = d3.scaleLinear() //name the values from 0 to 20 and make their values from .1-.7
        .domain([0, numOfLines])
        .range([0.8, 0.085])

    var yPos = d3.scaleLinear() //scalling for creating horizontal lines
        .domain([0, numOfLines])
        .range([0, 4])

    var x = d3.scaleLinear()
        .domain([0, 5])
        .range([0, width]);

    var y = d3.scaleLinear()
        .domain([0, 4.5])
        .range([height, 0]);

    // The line functions:
    var logistic = d3.range(numOfLines).map(function(i) {
        var odd = true
        var toReturn = xs.map(function(num) {
            return {
                "x": num,
                "y": logistic(num, thetaMap(i), i)
            }
        })
        return toReturn;
    })

    //make a greeting message for after the line animation.
    function writeGreeting(){

        var text_scale = width < 500 ? 1 : 0.65;
        console.log(text_scale);
        var intro_text = "Live Free or Dichotomize";

        var title = svg.append("text")
            .attr("font-size", 30)
            .attr("font-family", "optima")
            .attr("text-anchor", "end")
            .attr("fill-opacity", 0.65)
            .attr("x", width < 500 ? x(4.8) : x(4.2))
            .attr("y", width < 500 ? y(3.12) : y(2.1) )
            .attr("opacity", 0)
            .text(intro_text)
            .attr("font-size", function(d) { return ( text_scale*width - 8) / this.getComputedTextLength() * 24 + "px"; })
            .text("")
            .attr("opacity", 1);

        (function drawGreeting (i, start, greeting) {
            setTimeout(function () {

                //add next letter to the greeting in progress
                start += greeting[i];

                title.html(start) //append this to the html

                if (start.length < greeting.length) { //if the in progress greeting is less than the full, keep going.
                    drawGreeting(i+1,start,greeting);      //  increment i and call again.
                };
            }, 100)
        })(0, "", intro_text);
    }


    // The d3 stuff
    var line = d3.line()
        // .interpolate("basis")
        .x(function(d) { return x(d.x); })
        .y(function(d) { return y(d.y); });


    svg.selectAll(".line")
        .data(logistic)
        .enter().append("path")
        .attr("class", "line")
        .attr("id" , function(d, i){ return "line" + i;})
        .attr("d", line)
        .style("stroke-width", 2)
        .style("stroke", function(d, i) { return colors[i % 10] })
        .style("opacity", 0)
        .style("fill", "none")
        .on("mouseover", function(d){
            d3.select(this).style("stroke", "black")
        })
        .on("mouseout", function(d,i){
            d3.select(this).style("stroke", colors[i % 10])
        })

    // var introMessage = isMobile ? "tap" : "click"
    // var introMessage = "click"
    //
    // var intro = svg.append("text")
    //     .text(introMessage)
    //     .attr("font-size", 35)
    //     .attr("font-family", "optima")
    //     .attr("text-anchor", "middle")
    //     .attr("x", x(2.5))
    //     .attr("y", y(2.01))

    //kick it off on a click. (or tap)
    animatelines(2)
    // d3.select("svg").on("click", function() { animatelines(2) })

})

</script>






<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title"></span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml" rel="" target=""><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">PheWAS-ME, an app for exploration of multimorbidity patterns in PheWAS</h1>
                  <div>
        <div class="description">
          This post is a longer-form and less-formal accompaniment to the manuscript “PheWAS-ME: A web-app for interactive exploration of multimorbidity patterns in PheWAS” and accompanying application. As the first of three papers that make up my PhD dissertation, the project represents a significant collaborative effort bringing together Electronic Health Records (EHR) and Biobank data using R and Shiny.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">R</div>
                <div class="quarto-category">shiny</div>
                <div class="quarto-category">javascript</div>
                <div class="quarto-category">d3</div>
                <div class="quarto-category">EHR</div>
                <div class="quarto-category">biobank</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Nick Strayer </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">October 16, 2019</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<p>This post is a longer-form and less-formal accompaniment to the manuscript just up on MedRxiv: <a href="https://www.medrxiv.org/content/10.1101/19009480v1"><em>“PheWAS-ME: A web-app for interactive exploration of multimorbidity patterns in PheWAS”</em></a> and accompanying application.</p>
<blockquote class="twitter-tweet blockquote">
<p lang="en" dir="ltr">
PheWAS-ME: A web-app for interactive exploration of multimorbidity patterns in PheWAS <a href="https://t.co/H7htmgWsK0">https://t.co/H7htmgWsK0</a> <a href="https://twitter.com/hashtag/medRxiv?src=hash&amp;ref_src=twsrc%5Etfw">#medRxiv</a>
</p>
— medRxiv (<span class="citation" data-cites="medrxivpreprint">@medrxivpreprint</span>) <a href="https://twitter.com/medrxivpreprint/status/1187468124448145409?ref_src=twsrc%5Etfw">October 24, 2019</a>
</blockquote>
<script async="" src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
<p>As the first of three papers that make up my PhD dissertation, the project represents a significant collaborative effort bringing together Electronic Health Records (EHR) and Biobank data using R and Shiny.</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">The main dashboard view of Phewas-ME.</span> <img src="../../media/phewas_me/whole_app.png" class="img-fluid"></p>
<p>I have organized the following into three primary sections: <strong>what</strong> the app does, <strong>why</strong> it does it, and <strong>how</strong> it works.</p>
<section id="what" class="level1">
<h1>What</h1>
<section id="what-is-the-app" class="level2">
<h2 class="anchored" data-anchor-id="what-is-the-app">What is the app?</h2>
<p>The paper describes Phewas-Multimorbidity Explorer (Phewas-ME): a Shiny app that we have been producing over the past two years in collaboration with the Vanderbilt Drug Repurposing team. It is an interactive data exploration tool for digging into PheWAS results and the subject-level data that generated those results.</p>
</section>
<section id="what-is-phewas" class="level2">
<h2 class="anchored" data-anchor-id="what-is-phewas">What is Phewas?</h2>
<p>PheWAS is a statistical method for finding associations between a given genetic mutation (often a Single Nucleotide Polymorphism, or SNP) and phenotypes. It is a sibling to the GWAS (genome-wide association study). However, in GWAS, you fix your desired phenotype and scan the genome for associations, whereas in PheWAS, you fix a genetic mutation, and scan the ‘phenome.’</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">Slide from my talk <a href="http://nickstrayer.me/biostat_seminar/">Taking a network view of EHR and Biobank data to find explainable multivariate patterns.</a></span> <img src="../../media/phewas_me/phewas_vs_gwas.png" class="img-fluid"></p>
</section>
<section id="what-is-the-data" class="level2">
<h2 class="anchored" data-anchor-id="what-is-the-data">What is the data?</h2>
<p>Both PheWAS analyses and Phewas-ME take two types of subject-level data. First is information on a given genetic mutation. In this case, 0, 1, or 2 copies of the minor allele of an SNP. The second is the subject’s ‘phenome.’ In this case, a list of every phenotype that they had. In the paper and this blog post, we have used the <a href="https://phewascatalog.org/phecodes">Vanderbilt developed ‘Phecode,’</a> but any binary phenotype information works. <label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">In this post, when I refer to ‘codes’ I mean ’phecodes.</span></p>
</section>
</section>
<section id="why" class="level1">
<h1>Why</h1>
<section id="why-is-the-app-needed" class="level2">
<h2 class="anchored" data-anchor-id="why-is-the-app-needed">Why is the app needed?</h2>
<p>Phewas-ME came out of the need for subject-experts to dig through the very high-dimensional results of PheWAS analyses. Unlike a typical statistical analysis, which may return just a handful of p-values or effect-sizes of interest, a PheWAS analysis returns a single p-value and effect-size for every phenotype in the scanned phenome. In the case of Phecodes, this is around 1,500 (and goes into the many thousands for more traditional ICD-based encodings).</p>
<section id="previous-methods" class="level3">
<h3 class="anchored" data-anchor-id="previous-methods">Previous methods</h3>
<p>In the past, communicating these results was accomplished using two different tools.</p>
<p>First is a manhattan plot: or simply a plot of every phenotype investigated and their significance (as encoded by the negative log of the p-value). This plot allows the reader to pick out which codes are significant in the results.</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">Manhattan plot for PheWAS results. Taken from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5480096/pdf/nihms864620.pdf">Denny et al, 2016</a></span> <img src="../../media/phewas_me/basic_manhattan_plot.png" class="img-fluid"></p>
<p>The second is the ever-present table. A PheWAS table has columns on all sorts of values of interest such as p-value, effect-size, phenotype description, etc.. Due to a large number of phenotypes typically present, these tables often get filtered to only includes codes that fit some criteria, such as significance level.</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">Table of PheWAS results from simulated data provided with app.</span> <img src="../../media/phewas_me/phewas_results_table.png" class="img-fluid"></p>
</section>
<section id="problems-with-previous-methods" class="level3">
<h3 class="anchored" data-anchor-id="problems-with-previous-methods">Problems with previous methods</h3>
<p>While looking at the most significantly associated phenotypes is straightforward, it can mask important aspects of the data. Different phenotypes across the phenome correlate with each other in complex ways. These correlations manifest themselves in common patterns of phenotypes called multimorbidities (also referred to as comorbidities). Since PheWAS looks at single SNP and Phenotype associations, these correlations can be extremely hard to discover and reason about with a traditional manhattan plot and table report.</p>
</section>
</section>
<section id="why-does-our-app-fix-these-problems" class="level2">
<h2 class="anchored" data-anchor-id="why-does-our-app-fix-these-problems">Why does our app fix these problems?</h2>
<p>PheWAS-Multimorbidity Explorer helps subject-matter experts parse through the results of a PheWAS analysis by letting them isolate and explore specific regions of phenotypes by looking not only at their p-values and effects-sizes but also by the subject-level data that generated those results.</p>
<p>The analyst selects a set of phenotypes to explore by dragging a selection box around a region in a manhattan plot, choosing directly from a table, or using text-search. Once a desired set of phenotypes is selected, the application displays the subject-level data using an interactive force-directed network plot and an upset plot.</p>
<p><img src="../../media/phewas_me/subject_level_plots.png" class="img-fluid"></p>
<p>These subject-level visualizations allow the analyst to see patterns of comorbid phenotypes directly and to interrogate their potential causes. For instance, often, a group of codes are highly correlated because they are more or less specific definitions of the same phenotype, e.g., cancer -&gt; lung cancer -&gt; stage 2 lung cancer. <label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">In the above plot, you can see a cluster of phenotypes in light green. This cluster is caused by the afformentioned phenotype hierarchy, with each refering to the same broad-level phenotype with differing levels of specificity.</span> Another possibility is multimorbidity caused by drug side-effects; e.g., patients taking a drug to treat the cancer phenotype likely also have the nausea phenotype.</p>
</section>
</section>
<section id="how" class="level1">
<h1>How</h1>
<section id="how-is-interaction-used-in-the-app" class="level2">
<h2 class="anchored" data-anchor-id="how-is-interaction-used-in-the-app">How is interaction used in the app?</h2>
<p>Interaction with Phewas-ME centers around one primary thing: a list of currently selected phenotypes. The typical way to choose these phenotypes is by using the Phewas results panel.</p>
<p><img src="../../media/phewas_me/phewas_results_panel.png" class="img-fluid"></p>
<p>This panel includes the traditional manhattan plot and results table; however, they are now interactive. Selecting codes using this panel can be done in a few different ways.</p>
<div class="cell">
<style type="text/css">
.giphy-holder {
  max-width: 95vw;
  width:100%;
  height:0;
  padding-bottom:85%;
  position:relative;
}
</style>
</div>
<section id="code-selection" class="level3">
<h3 class="anchored" data-anchor-id="code-selection">Code Selection</h3>
<section id="region-dragging" class="level4">
<h4 class="anchored" data-anchor-id="region-dragging">Region Dragging</h4>
<p>The user can drag a box around a region in the manhattan plot, selecting the codes within. Adding another region to the selection is accomplished by holding down the ‘a’ key (for add) and dragging another box. Conversely, removing a region of codes from the selection can be done by holding down the “d” key and dragging a box.</p>
<iframe src="https://giphy.com/embed/kHqnsHdw9a7G40AzVo" width="480" height="348" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/kHqnsHdw9a7G40AzVo">via GIPHY</a>
</p>
</section>
<section id="code-clicking" class="level4">
<h4 class="anchored" data-anchor-id="code-clicking">Code Clicking</h4>
<p>Clicking individual phenotypes in either the plot or the table toggles their selected-ness. This slower but more precise method allows for fine-grained tuning of the selected codes.</p>
<iframe src="https://giphy.com/embed/VEVOfKPJ83gMfsc5XI" width="480" height="348" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/VEVOfKPJ83gMfsc5XI">via GIPHY</a>
</p>
</section>
<section id="search" class="level4">
<h4 class="anchored" data-anchor-id="search">Search</h4>
<p>Above the results table is a search bar. The user can search for phenotypes by name or description. Codes that match a supplied search query are raised to the top of the results table and highlighted. The user can then select codes as needed.</p>
<iframe src="https://giphy.com/embed/dAQTSX8ZHpGRWFg4EJ" width="404" height="480" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/dAQTSX8ZHpGRWFg4EJ">via GIPHY</a>
</p>
</section>
<section id="network-filtering" class="level4">
<h4 class="anchored" data-anchor-id="network-filtering">Network Filtering</h4>
<p>In the subject-level network plot, nodes corresponding to phenotypes can be used to filter the selection further. After selecting codes in the network plot by clicking, users can delete from the current selection, isolate a subset, or invert (this flips the definition of a connection to a phenotype in the network to the lack of that phenotype in a subject’s phenome).</p>
<iframe src="https://giphy.com/embed/lnxeLWOIB35BJWOif0" width="480" height="324" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/lnxeLWOIB35BJWOif0">via GIPHY</a>
</p>
</section>
</section>
<section id="other-interaction" class="level3">
<h3 class="anchored" data-anchor-id="other-interaction">Other interaction</h3>
<p>Outside of phenotype selection, there are a few other forms of interaction.</p>
<section id="pattern-highlighting" class="level4">
<h4 class="anchored" data-anchor-id="pattern-highlighting">Pattern Highlighting</h4>
<p>When a user clicks on a given pattern in the Upset Plot panel, the network plot updates to highlight the subjects who have that pattern. Similarly, if a phenotype node in the network plot itself is moused over, the subjects connected to that phenotype are highlighted.</p>
<iframe src="https://giphy.com/embed/Tf95vix9K3HmhjjvFF" width="444" height="480" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/Tf95vix9K3HmhjjvFF">via GIPHY</a>
</p>
</section>
<section id="info-tooltips" class="level4">
<h4 class="anchored" data-anchor-id="info-tooltips">Info Tooltips</h4>
<p>In both the manhattan plot and the network plot, when a given phenotype’s point is moused-over, a tooltip with all the supplied information for that phenotype appears.</p>
</section>
<section id="snp-filtering" class="level4">
<h4 class="anchored" data-anchor-id="snp-filtering">SNP Filtering</h4>
<p>A check-box by the network plot allows the user to show only carriers of the minor allele of interest in the network plot. This helps differentiate between genetics driven network patterns and simply population trends.</p>
<iframe src="https://giphy.com/embed/dWxVaK7R0wxEw3yaUN" width="480" height="328" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/dWxVaK7R0wxEw3yaUN">via GIPHY</a>
</p>
</section>
<section id="upset-interaction" class="level4">
<h4 class="anchored" data-anchor-id="upset-interaction">Upset Interaction</h4>
<p>There are two helpful interaction aides not present in the original UpSet plot.</p>
<p><strong>Singleton Toggle</strong> First, a toggle allows the hiding of single-phenotype patterns (i.e., subjects who had just a single phenotype). Often these singletons can crowd out the more exciting comorbidity patterns, so by hiding them, the user can focus on the patterns.</p>
<iframe src="https://giphy.com/embed/MXcJgYiYrCQVVDyJQu" width="480" height="328" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/MXcJgYiYrCQVVDyJQu">via GIPHY</a>
</p>
<p><strong>Minimum Pattern Size Slider</strong> Second, a slider allows the user to set the threshold for inclusion in the plot based on the number of times a pattern appears. Often there will be patterns of phenotypes that are only had by one or two subjects. Because of their tiny sample-sizes, these are usually not of interest, so filtering them out can again improve the plot’s effectiveness.</p>
</section>
</section>
</section>
<section id="how-was-it-made" class="level2">
<h2 class="anchored" data-anchor-id="how-was-it-made">How was it made?</h2>
<p>There is a lot of exciting tech going on behind the scenes of Phewas-ME.</p>
<section id="general-app" class="level3">
<h3 class="anchored" data-anchor-id="general-app">General App</h3>
<section id="framework" class="level4">
<h4 class="anchored" data-anchor-id="framework">Framework</h4>
<p>The main app is built using the Shiny package/framework in R. An R package was created to wrap all the code, easing deployment and customization. There are two main functions in the package: <code>run_me()</code> which takes dataframes of all necessary information as input and starts at the main dashboard; and also <code>build_me_app()</code> which returns an app that starts at a data-loading screen where the user can either load data using spreadsheets, or pick from a list of pre-populated datasets. All individual plots and panels exported by the package as shiny modules. This modular format eases the creation of customized versions of PheWAS-ME.</p>
</section>
<section id="state-management" class="level4">
<h4 class="anchored" data-anchor-id="state-management">State Management</h4>
<p>The reactivity protocol of shiny works great if you have a single component of your app that listens to and modifies unique parts of your apps state. In PheWAS-ME, multiple components listen to the same state variables (for instance, both the PheWAS results table and the network plot modify the currently selected codes). A state management system inspired by Javascript’s Redux was built to keep each component pure and modular.</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">Broad level idea of how Redux works, via the <a href="https://blog.jscrambler.com/becoming-familiar-with-redux/">jscrabler blog</a></span> <img src="https://css-tricks.com/wp-content/uploads/2016/03/redux-article-3-03.svg" class="img-fluid"></p>
<p>In this system, a single message-passing reactive value is supplied to all of the separate components. When a component needs to modify the state, it sends a message through that reactive value to the main app.</p>
<p>As an example, if the user removes a code using the network plot. The network plot module sends a list to the message passing reactive variable containing an action type: <code>'delete code'</code>, and then a payload: <code>'008.12'</code>. Then at the main app level, an <code>observeEvent()</code> chunk that watches that message passing reactive reads the message and modifies the appropriate state variables, in this case, the selected codes list.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>shiny<span class="sc">::</span><span class="fu">observeEvent</span>(<span class="fu">app_interaction</span>(),{</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  action_type <span class="ot">&lt;-</span> <span class="fu">app_interaction</span>() <span class="sc">%&gt;%</span> <span class="fu">pluck</span>(<span class="st">'type'</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  action_payload <span class="ot">&lt;-</span> <span class="fu">app_interaction</span>() <span class="sc">%&gt;%</span> <span class="fu">pluck</span>(<span class="st">'payload'</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  action_type <span class="sc">%&gt;%</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">switch</span>(</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">delete =</span> {</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>        codes_to_delete <span class="ot">&lt;-</span> <span class="fu">extract_codes</span>(action_payload)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>        prev_selected_codes <span class="ot">&lt;-</span> state<span class="sc">$</span><span class="fu">selected_codes</span>()</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>        state<span class="sc">$</span><span class="fu">selected_codes</span>(</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>          prev_selected_codes[<span class="sc">!</span>(prev_selected_codes <span class="sc">%in%</span> codes_to_delete)]</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>      },</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">selection =</span> {</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>        ...</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>      },</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>      ...</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By isolating all state modification in a single observable chunk, the primary app state is only ever modified in a single place. This isolation reduces the amount of overhead required to reason about state changes, both for Shiny and the programmer.</p>
</section>
</section>
<section id="r-to-javascript-communications" class="level3">
<h3 class="anchored" data-anchor-id="r-to-javascript-communications">R to Javascript Communications</h3>
<p>Every visual in the app is built using custom javascript and HTML. The fantastic package <a href="https://rstudio.github.io/r2d3/"><code>R2D3</code></a> is used to facilitate the handoff of data between R/Shiny and the javascript visualization code.</p>
<section id="manhattan-plot" class="level4">
<h4 class="anchored" data-anchor-id="manhattan-plot">Manhattan Plot</h4>
<p>The manhattan plot is built using <a href="https://d3js.org">d3js</a>, the popular javascript visualization library. With such a large number of points plotted, finding points within a selected area can be very slow when naively searching through every point. <label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/8b/Point_quadtree.svg/600px-Point_quadtree.svg.png" class="img-fluid"> A 2d quadree storing datapoints. These subdividing boxes are stored as in a tree-structure, which is always a big-O help. <a href="https://en.wikipedia.org/wiki/Quadtree">From wikipedia</a></span> The javascript code utilizes a special data structure known as a <a href="https://en.m.wikipedia.org/wiki/Quadtree">quad-tree</a> to speed up finding codes within a region. By storing data in a hierarchical location-aware tree, a quad-tree only searches through points near the selection, cutting down on the computation needed and speeding up app reaction times.</p>
</section>
<section id="phewas-table" class="level4">
<h4 class="anchored" data-anchor-id="phewas-table">PheWAS Table</h4>
<p>Like the manhattan plot, the PheWAS table has tricks to help keep the app responsive while also showing large amounts of data. The browser stores all the elements on a webpage in a text-based format called the DOM (or Document Object Model). (What you get when you right-click and say ‘view source.’) When a browser renders the page, it parses through the DOM and figures out how to place that element on the screen. Because the browser has to parse through so much every time it wants to update the screen, there are limits to how many elements can be stored in the DOM before things start slowing down. <label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote"><span class="lightbox"> <img src="../../media/phewas_me/optimal_dom_size.png" class="img-fluid"> Recomendations from <a href="https://developers.google.com/web/tools/lighthouse/audits/dom-size">developers.google.com</a> for optimal DOM performance. Oops! </span></span>Displaying a table of more than 1,500 phenotypes, along with 5+ columns, quickly hits that limit.</p>
<p>The trick used to display a large table while not overloading the DOM is only rendering a small portion of the table at a time. At any moment, only 50 rows of the table are actually in the DOM. As the user scrolls down, the javascript writes new rows at the bottom and removes old rows from the top. As to maintain the illusion that the entire table is available for scrolling, secret elements placed above and below the visible rows that expand and contract to simulate a table with every row filled in. The user sees a table just like it was full of 1,000 rows, but the browser only has to render 50. Implementing this optimization resulted in interaction latency dropping by ~4 times.</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">Performance improvements in results table after implementing smart-table. Via the Chrome devtool’s <a href="https://developers.google.com/web/tools/lighthouse/">lighthose audits tool.</a></span> <img src="../../media/phewas_me/table_performance_improvements.png" class="img-fluid"></p>
</section>
<section id="upset-plot" class="level4">
<h4 class="anchored" data-anchor-id="upset-plot">UpSet Plot</h4>
<p>Like the manhattan plot, the upset plot is drawn using d3. Unlike the results table, there are not too many elements to draw here, so no fancy tricks are needed to speed it up.</p>
</section>
<section id="network-plot" class="level4">
<h4 class="anchored" data-anchor-id="network-plot">Network Plot</h4>
<p>The network plot was where the most coding and optimization hours were spent. The goal was to have a plot that could show every subject’s data. To draw every subject often involved drawing tens of thousands of points and even more links. Also, since the user would be on-the-fly reconfiguring the data displayed, we needed to calculate the layout in real-time.</p>
<p><strong>Drawing lots of points and lines</strong> Plotting lots of points and lines had a relatively simple solution: use the Canvas element rather than the SVG. <a href="https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API">The Canvas element</a> allows the user to draw to the screen at the pixel-level, basically coding an image. Because the Canvas element acts as a fancy image, the DOM only has a single item added is not slowed down.</p>
<p><strong>Adding interactivity</strong> The only issue with the canvas element: there’s no easy way to detect if a component of the chart is interacted with by the user. We wanted interactivity with the phenotypes (e.g., mouseover and selection). Adding interactivity was solved by overlaying an SVG element on top of the Canvas element. The SVG element draws only the phenotype nodes. We could then use the SVG’s built-in intersection observers to detect interaction, rather than building a custom solution using things like <a href="https://observablehq.com/@camargo/canvas-picking-with-a-hidden-canvas">hidden colors.</a></p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">Both the Canvas and the SVG are absolutely positioned within the parent to make sure they overlap properly.</span> <img src="../../media/phewas_me/network_canvas_and_svg.png" class="img-fluid"></p>
<p><strong>Exporting paper-ready plots</strong> One feature that was frequently requested was the ability to export the network plot for use in a publication. Since canvas elements are just raster-images, they didn’t scale well when exported. To allow for clean high-res exports, an ‘export mode’ was added that switches all rendering to the SVG (which is vector-based) and adds a download button. Also available is a callouts mode where little draggable annotations for each phenotype name are added. When a user presses the download button, the SVG element is converted to the SVG file format and downloaded the user’s computer. Adobe Illustrator or other vector-based tools can be used to insert the plot into a figure or PDF.</p>
<iframe src="https://giphy.com/embed/H3edojKWcvr3fY6tDX" width="480" height="338" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/H3edojKWcvr3fY6tDX">via GIPHY</a>
</p>
<p><em>Not visible due to compression for video: a slight increase in sharpness when switching to all SVG-Rendering.</em></p>
<p><strong>Calculating a force-layout in real time</strong> Calculating the layout was relatively easy to do: D3 includes a sub-library for running force-layout simulations. The problem is these force simulations are computationally intensive, especially when you start running them on large networks. When you run computationally intensive tasks in javascript, your app will slow to a crawl because the same thread responsible for updating the screen is also running those calculations.</p>
<p>Our solution was to offload the layout calculation to another thread using a technology called ‘web workers.’ <a href="https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers">Web workers</a> allow javascript code to run on another thread and communicate with the main thread through a series of asynchronous callbacks.</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote"><img src="https://www.barbarianmeetscoding.com/images/barbaric-basics-web-workers.jpg" class="img-fluid"> Web workers are there own threads with their own event loops, so they don’t slow down your main page. Source: <a href="https://www.barbarianmeetscoding.com/blog/2015/02/13/barbaric-basics-web-workers">Barbaric Basics.</a></span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">// Function to send new network data to webworker</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">send_new_job</span>(network_data<span class="op">,</span> callbacks){</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="kw">const</span> {on_progress_report<span class="op">,</span> on_finish<span class="op">,</span> on_layout_data} <span class="op">=</span> callbacks<span class="op">;</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="co">// Initialize the worker</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  worker <span class="op">=</span> <span class="kw">new</span> <span class="bu">Worker</span>(worker_url)<span class="op">;</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">// Send worker the data we are working with</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  worker<span class="op">.</span><span class="fu">postMessage</span>(network_data)<span class="op">;</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">// Control what is done when a message is received from the webworker</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  worker<span class="op">.</span><span class="at">onmessage</span> <span class="op">=</span> <span class="kw">function</span>(<span class="bu">event</span>) {</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">switch</span> (<span class="bu">event</span><span class="op">.</span><span class="at">data</span><span class="op">.</span><span class="at">type</span>) {</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>      <span class="cf">case</span> <span class="st">"progress_report"</span><span class="op">:</span> <span class="cf">return</span> <span class="fu">on_progress_report</span>(<span class="bu">event</span><span class="op">.</span><span class="at">data</span><span class="op">.</span><span class="at">progress</span>)<span class="op">;</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>      <span class="cf">case</span> <span class="st">"layout_data"</span><span class="op">:</span> <span class="cf">return</span> <span class="fu">on_layout_data</span>(<span class="bu">event</span><span class="op">.</span><span class="at">data</span>)<span class="op">;</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>      <span class="cf">case</span> <span class="st">"end"</span><span class="op">:</span> <span class="cf">return</span> <span class="fu">on_finish</span>(<span class="bu">event</span><span class="op">.</span><span class="at">data</span>)<span class="op">;</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>  }<span class="op">;</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>}<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When the network plot receives new data, it spins up one of these web-worker threads, sends the data to it to run the layout simulation. The simulation web worker sends back occasional updates on the node positions to the main thread to update the plot, assuring the user it is doing something and not frozen. This adaptation meant that massive networks could be laid out and rendered without causing any perceivable slowdowns on the main app.</p>
<iframe src="https://giphy.com/embed/RGe5NHijYCuZm2R25D" width="480" height="340" frameborder="0" class="giphy-embed" allowfullscreen="">
</iframe>
<p>
<a href="https://giphy.com/gifs/RGe5NHijYCuZm2R25D">via GIPHY</a>
</p>
</section>
</section>
</section>
<section id="how-can-i-use-the-app-myself" class="level2">
<h2 class="anchored" data-anchor-id="how-can-i-use-the-app-myself">How can I use the app myself?</h2>
<p>Do you have data of the form described and want to try the app out yourself? There are two main methods for doing so.</p>
<section id="hosted-version" class="level3">
<h3 class="anchored" data-anchor-id="hosted-version">Hosted Version</h3>
<p>We have a version of Phewas-ME up and <a href="https://prod.tbilab.org/phewas_me/">running on our lab’s website</a> that anyone can use. There is a simulated dataset for exploration, but a data loading screen allows you to test out with data you upload yourself.</p>
</section>
<section id="an-r-package" class="level3">
<h3 class="anchored" data-anchor-id="an-r-package">An R Package</h3>
<p>As mentioned before, the application lives in an R package. Because of this, running it locally is as easy as installing the package from GitHub using…</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">'tbilab/meToolkit'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>…and then starting an app up with the function <code>run_me()</code>.</p>
<p>If you want to learn how to make an app that loads directly to a given dataset or preload data, check out the online documentation for the package at <a href="https://prod.tbilab.org/phewas_me_manual/">prod.tbilab.org/phewas_me_manual</a>.</p>
<p><label for="tufte-mn-" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-" class="margin-toggle"><span class="marginnote">The package’s beautiful website, courtesy of <a href="https://pkgdown.r-lib.org/">pkgdown!</a></span> <img src="../../media/phewas_me/package_site.png" class="img-fluid"></p>
</section>
</section>
</section>
<section id="question-or-comments" class="level1">
<h1>Question or Comments?</h1>
<p>If you would like to know any more about the application, package, or manuscript, send me an email or twitter DM!</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://utteranc.es/client.js" repo="lfod/lfod.github.io" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script>
</div> <!-- /content -->



</body></html>